plugins/inputs/fritzbox/fritzbox.go

```diff
@@ -0,0 +1,402 @@
+	}
+	if choice.Contains("ppp", plugin.Collect) {
+		plugin.serviceHandlers[wanpppconn.ServiceShortType] = plugin.gatherPppInfo
+	}
+	if choice.Contains("dsl", plugin.Collect) {
+		plugin.serviceHandlers[wandslifconfig.ServiceShortType] = plugin.gatherDslInfo
+	}
+	if choice.Contains("wlan", plugin.Collect) {
+		plugin.serviceHandlers[wlanconfig.ServiceShortType] = plugin.gatherWlanInfo
+	}
+	if choice.Contains("hosts", plugin.Collect) {
+		plugin.serviceHandlers[hosts.ServiceShortType] = plugin.gatherHostsInfo
+	}
+}
+
+func (plugin *Fritzbox) Gather(acc telegraf.Accumulator) error {
+	for _, deviceClient := range plugin.deviceClients {
+		plugin.gatherDevice(acc, deviceClient)
+	}
+	return nil
+}
+
+func (plugin *Fritzbox) gatherDevice(acc telegraf.Accumulator, deviceClient *tr064.Client) {
+	plugin.Log.Debugf("Querying %s", deviceClient.DeviceUrl.Redacted())
+	services, err := deviceClient.Services(tr064.DefaultServiceSpec)
+	if err != nil {
+		acc.AddError(err)
+		return
+	}
+	for _, service := range services {
+		serviceHandler := plugin.serviceHandlers[service.ShortType()]
+		if serviceHandler == nil {
+			continue
+		}
+		err := serviceHandler(acc, deviceClient, service)
+		if err != nil {
+			acc.AddError(err)
+			return
+		}
+	}
+}
+
+type serviceHandlerFunc func(telegraf.Accumulator, *tr064.Client, tr064.ServiceDescriptor) error
+
+func (plugin *Fritzbox) gatherDeviceInfo(acc telegraf.Accumulator, deviceClient *tr064.Client, service tr064.ServiceDescriptor) error {
+	serviceClient := deviceinfo.ServiceClient{
+		TR064Client: deviceClient,
+		Service:     service,
+	}
+	info := &deviceinfo.GetInfoResponse{}
+	err := serviceClient.GetInfo(info)
+	if err != nil {
+		return err
+	}
+	tags := make(map[string]string)
+	tags["source"] = serviceClient.TR064Client.DeviceUrl.Hostname()
+	tags["service"] = serviceClient.Service.ShortId()
+	fields := make(map[string]interface{})
+	fields["uptime"] = info.NewUpTime
+	fields["model_name"] = info.NewModelName
+	fields["serial_number"] = info.NewSerialNumber
+	fields["hardware_version"] = info.NewHardwareVersion
+	fields["software_version"] = info.NewSoftwareVersion
+	acc.AddFields("fritzbox_device", fields, tags)
+	return nil
+}
+
+func (plugin *Fritzbox) gatherWanInfo(acc telegraf.Accumulator, deviceClient *tr064.Client, service tr064.ServiceDescriptor) error {
+	serviceClient := wancommonifconfig.ServiceClient{
+		TR064Client: deviceClient,
+		Service:     service,
+	}
+	commonLinkProperties := &wancommonifconfig.GetCommonLinkPropertiesResponse{}
+	err := serviceClient.GetCommonLinkProperties(commonLinkProperties)
+	if err != nil {
+		return err
+	}
+	// Use public IGD service instead of the WANCommonInfo one for total bytes stats, because IGD supports uint64 counters
+	igdServices, err := deviceClient.ServicesByType(tr064.IgdServiceSpec, "WANCommonInterfaceConfig")
+	if err != nil {
+		return err
+	}
+	if len(igdServices) == 0 {
+		return fmt.Errorf("IGD service not available")
+	}
```

Comment:
Would it make sense to call/process `WANCommonInfo` instead in this scenario?

Reply:
Implemented via 8976c29fe4ca62dbfa2acf2a4e32a0e48c436a2a